Allow for tooltips with rich content

omarkhan opened the pull request:
This pull request implements tooltips with rich content, including links.

**Screenshot:**

<img width="835" alt="screen shot 2016-03-20 at 09 55 48" src="https://cloud.githubusercontent.com/assets/793379/13902626/2b7fdd9c-ee82-11e5-9a4c-4f7ba448508d.png">

**Sandbox URL:** http://pr11854.sandbox.opencraft.com/courses/course-v1:edX+DemoX+Demo_Course/courseware/d8a6192ade314473a78242dfeedfbf5b/4828664659a64c43812b52fec58ee3fc/1 (persistent database)

**Partner Information:** 3rd party-hosted open edX instance

**Implementation:** The current tooltip implementation has three limitations that prevent it from being used for rich content:

- Tooltips follow the mouse, so it is effectively impossible to click on any links in the tooltips
- If the content is too wide, tooltips can overflow off the left side of the viewport
- Focusable elements within the tooltip cannot be focused using the keyboard

This pull request solves these problems by:

- Making tooltips static relative to the position of the parent element
- Preventing tooltips from going past the left edge of the display
- Making tooltips inline with the parent element, so that tooltip content receives focus in the correct order

**Usage:** The same as the existing tooltips. Add a `data-tooltip` attribute to any html element.

**Accessibility:** This pull request adds a focus handler to tooltip elements so that the tooltip is displayed when the element gets focus using the keyboard. Focusable elements within the tooltip can also be accessed using the keyboard.

**Settings**
```yaml
edx_ansible_source_repo: https://github.com/open-craft/configuration.git
configuration_version: ekolpakov/celery-with-certs
```

edx-webhook commented:
Thanks for the pull request, @omarkhan! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-platform&number=11854

omarkhan commented:
jenkins run bokchoy

omarkhan commented:
@Kelketek some bokchoy tests are failing, but for unrelated features. Could be flaky tests, I'm running them again to check

Kelketek commented:
@omarkhan :+1: when the tests pass. Though let me know if that requires any significant changes.

omarkhan commented:
@Kelketek to fix the bok choy test failures I had to remove the test code that disables CSS transforms: 84001af. My tooltip changes use CSS transforms to keep the tooltip above the element.

Let's see if that breaks any of the other tests...

Kelketek commented:
:+1:

omarkhan commented:
Commits squashed

edx-webhook commented:
Thanks for the pull request, @omarkhan! I've created [OSPR-1205](https://openedx.atlassian.net/browse/OSPR-1205) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

cahrens commented:
@omarkhan has an accessibility review been done of this? Can a keyboard user move focus into the tooltip and navigate to press the links?

@jbarciauskas I definitely think this PR should have a UX review. FYI @cptvitamin and @clrux

omarkhan commented:
@cahrens I don't think anyone has done a formal accessibility review. Keyboard users can open the tooltips by moving focus to the triggering element but focus does not move inside the tooltips.

cahrens commented:
@OmarIthawi If the rich tooltip can only be used with the mouse, I don't think it meets our accessibility requirements. @cptvitamin can you advise?

OmarIthawi commented:
> @OmarIthawi If content within the rich tooltip can only be accessed via the mouse, I don't think it meets our accessibility requirements. @cptvitamin can you advise?
 Show all checks


@cahrens I think you meant to tag @omarkhan instead of me :), no worries. I will just unsubscribe.

cptvitamin commented:
Thanks for pinging me @cahrens. @omarkhan this is looking pretty good.  I like how you added basic keyboard support.  If this feature didn't also support interactive content (links in this case) we would be almost there.  (You could use aria-live to force the screen reader to read the contents of the tooltip when focus was received.)  HOwever, because we support interactive content, we have to make sure that keyboard and the screen reader users can also access the content.

I put a little thought into how I would solve this issue, or build such a feature and this is how I would go about it:  Use ```<figure>``` and ```<figcaption>``` to contain the primary content and the tooltip.  The tooltip would be contained in the ```figcaption``` element (many people think that the ```figure``` element is reserved for images, but you can put anything you want in there.)  The figcaption can have an SR class (edx specific) on it (usually reserved for serving content only to screen readers).  This would give screen reader users access to the content, including the interactive content.  However, this breaks one of the cardinal rules of accessibility, that you should never put interactive content off screen.  That would make the visible keyboard focus disappear while anything offscreen had focus.  This is where the "tooltip" behavior kicks in.  Using javascript, detect whenever the figure element or any of its children has keyboard focus.  When this is true, toggle off the SR class, exposing the tooltip on the screen.  This has the added benefit of scoping both the primary content and the tooltip inside of one container.  The benefit here is that whenever the figure element has focus, the entire thing will scroll into view.  I noticed that when I reverse tabbed through the interface, the tooltip appeared above the image, but nothing above the image was in the viewport, so I did not see the tooltip.  If you add the tabindex=0 to the figure element, the content PLUS the tooltip will scroll into the viewport.

I also noted that the tooltip for the link stayed visible on the screen even after tabbing away (blur) from it (obscuring the content behind it).

Generally, the text for the tooltips was *very* small.  I would also verify that contrast requirements are met.  The blue link text on black is likely not compliant but I did not find an easy way to grab the color values to test.

On a side note, this example includes a hard accessibility failure not related to the feature.  The image example does not include an alt attribute, nor does the one that is included as a tooltip.  I also noted some improperly closed elements on the button element and redundant tabindex=0 on both the link and the button example (these are natively focusable elements, so specifying a tabindex is not necessary). :
```
<button data-tooltip="<img src='https://upload.wikimedia.org/wikipedia/commons/b/b1/Ewok_SWExhibition.jpg' width='250' height='241' ></button>" tabindex="0">Button</button>
``` 

Hope this helps.

omarkhan commented:
Thanks for the thorough review @cptvitamin! I think I have a pretty good idea of what needs to be done now. The main obstacle here is that in the current implementation all the tooltips are actually a single element directly under `<body>`, with the content swapped out dynamically. Am I right in assuming that the tooltips will have to be inline with the content so that elements within tooltips receive focus in the correct order?

CC @bradenmacdonald

cptvitamin commented:
@omarkhan that is exactly correct. You would want the tooltips to be adjacent to their triggering elements. Otherwise you would have to manually manage focus, which is not impossible to do but will require some effort. If you go with  any other solution than the one I recommended you will also have to solve the problem of informing the user that there is a tooltip. This is important on interactive controls like links or buttons (if the user doesn't know supplemental information *follows* the control, how would they know to consume it before activating the control? This will be difficult.

omarkhan commented:
jenkins run js

bradenmacdonald commented on cms/static/sass/_base.scss:
```diff
@@ -595,26 +595,27 @@ hr.divide {
+  > .tooltip-content {
+    @include font-size(12);
+    @extend %t-regular;
+    line-height: 26px !important;
```
Is there some way to avoid using `!important` here? Increase the specificity of the selector, perhaps?

omarkhan replied:
The issue here is that since the tooltip is now inline with the content, any styles that apply to the content will also apply to the tooltip. The `.xmodule_display.xmodule_HtmlModule *` style in particular has greater specificity, and gives a line-height of 1.4em. I am not sure how to fix this in the general case other that by applying the styles directly to the element in JS.

bradenmacdonald commented on lms/static/sass/course/base/_base.scss:
```diff
@@ -133,30 +133,39 @@ img {
+    padding: 0 10px;
+    border-radius: 3px;
+    background: rgba(0, 0, 0, 0.85);
+    color: $white;
```
Why is the same SASS code in three different files? Can we use a mixin?

omarkhan replied:
Done. I had to leave out some mixins that are only available in studio.

bradenmacdonald commented on common/test/acceptance/tests/helpers.py:
```diff
@@ -139,7 +139,7 @@ def enable_jquery_animations(page):
 def disable_css_animations(page):
     """
-    Disable CSS3 animations, transitions, transforms.
+    Disable CSS3 animations and transitions.
```
This seems like a change that would break other tests. Can you instead just use a parameter or decorator so that `disable_animations(page)` is not called for the specific tests you need?

Or, better yet: Instead of setting an absolute `top` position and using `transform: translateY(-100%)`, just leave `top` undefined and set an absolute `bottom` position. Yeah, do that.

omarkhan replied:
Good idea, done.

bradenmacdonald commented:
@omarkhan When I put a big image in the tooltip with the latest code, it appears like this (using placehold.it/100x500):
![screen shot 2016-05-06 at 12 28 57 pm](https://cloud.githubusercontent.com/assets/945577/15084102/2cd43c1e-1386-11e6-949d-bd757fc0d86f.png)

Also, please add the demo course back to the sandbox (maybe change it to use persistent DB?)

bradenmacdonald commented on .jshintrc:
```diff
@@ -115,6 +115,9 @@
         "_",
 
+        // Backbone.js globals
+        "Backbone",
```
This may be out of scope, but I think you're supposed to use requirejs, e.g.
```
require(["jquery", "backbone"], function($, Backbone) {
```
rather than assume the global is available. I know the existing code was not using requirejs and was already assuming a jquery global but you may as well clean it up rather than adding jshint exceptions.

omarkhan replied:
I tried this, and it works. The problem is, the 'common' js tests are not currently configured to use requirejs, and I get `TypeError: require is not a function` when I try to run them. I have tried including requirejs in the karma config, but I can't get it to work. I have reverted the change for now, I will try again if I have time at the end of the sprint.

bradenmacdonald commented on common/static/js/src/tooltip_manager.js:
```diff
@@ -1,109 +1,239 @@
-                'mouseout.TooltipManager': this.hideTooltip,
-                'click.TooltipManager': this.click
-            }, this.SELECTOR);
+        template: _.template('<div class="tooltip-content"><%= content %></div><div class="tooltip-arrow">▾</div>'),
```
Is there any advantage to having the tooltip arrow as a real element rather than a CSS pseudo-element?

omarkhan replied:
CSS pseudo-elements can't be styled or positioned, and we need to be able to move the arrow to the side if the tooltip position is offset by the edge of the viewport.

bradenmacdonald replied:
Not sure what you mean, as you can apply styles, including `position` to pseudo-elements. But I don't care which way you implement this; I was just curious.

omarkhan replied:
Sorry I was not being clear: pseudo-elements can't be positioned from javascript code. Last time I checked at least, things may have moved on since the last time I tried to do this! I would love to be proved wrong here

bradenmacdonald commented on common/static/js/src/tooltip_manager.js:
```diff
@@ -1,109 +1,239 @@
+            if (this.timer) {
+                clearTimeout(this.timer);
+            }
+            var remove = _.bind(this.remove, this);
```
Note (no change needed): edX doesn't support older browsers so it's fine to use native javascript `bind()` if you want, rather than the underscore version.

bradenmacdonald commented on common/static/sass/_tooltips.scss:
```diff
@@ -3,17 +3,26 @@
     color: $white;
+
+    a {
+      color: #66aed0 !important;
```
For colors, you should use one of [the color variables](https://github.com/edx/edx-platform/blob/5bb0e908/cms/static/sass/_variables.scss) instead of putting the RGB value.

omarkhan replied:
I tried that but the sass refused to compile. I'm not sure why, because the color variables are available in both the lms and studio sass. I will have another look.

bradenmacdonald replied:
Weird; you can see `$white` is being used just below.

omarkhan commented:
@cptvitamin this is ready for another look now. I have modified the tooltip code to put tooltips adjacent to the triggering element. This ensures that elements inside the tooltip can receive focus with the keyboard only, in the correct order. I have updated the sandbox, could you have a look and let me know if I'm on the right track? http://pr11854.sandbox.opencraft.com/courses/course-v1:edX+DemoX+Demo_Course/courseware/d8a6192ade314473a78242dfeedfbf5b/4828664659a64c43812b52fec58ee3fc/1

I have not implemented your suggestion of using `<figure>` and `<figcaption>`, mainly because that would require changing the markup everywhere that tooltips are currently used, which would be a big change.

The tooltips are currently only created when the triggering element gets focus. Will that be a problem for discoverability? The tricky thing with creating them before they are triggered is that we don't know when the triggering element is created. In studio, for example, the course content appears well after domready. Could we use something like [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) for this?

cptvitamin commented:
@omarkhan I have a fundamental issue with this feature, but I also have a solution that I think would work.  The tooltip content is designed to be consumed optionally.  There is value in that, otherwise we wouldn't be designing this feature.  Currently, the way this is implemented, a keyboard and/or screenreader user is forced to consume the content of the tooltip.  For them, it is not optional. This is not an equivalent experience.  To do this right, we would treat it more like a modal dialog box.  The modal would be invoked by a click, or hover (but not focus).  Since we anticipate using this feature with content that is already focusable/interactive, like links and buttons, we would need to add an icon adjacent to the element that has the associated tooltip.  The icon would visually indicate that a tooltip is available, which is an improvement over the current implementation (there is not visual cue as to what elements have tooltips and which don't).  The modal would appear when hovering over the triggering element OR the icon.  However, the tooltip would only appear for keyboard and screen reader users if they deliberately click on the icon.  Regular modal behaviors would kick in from there.  You can read more about this pattern here:

* https://www.w3.org/TR/wai-aria-practices/#dialog_tooltip
* https://www.w3.org/TR/wai-aria-practices/#dialog_modal

I also wanted to point out that the CSS generated "downward arrow" which is added in .tooltip-arrow needs to be aria-hidden=true.  this is accessible to screen reader users.

omarkhan commented:
@cptvitamin we have put together [a proposal for a new implementation](https://docs.google.com/document/d/1VQG2zlL45STxykVBqwGZb1SUkirYPt3iOJBajBmd9ew/edit?usp=sharing) that should meet accessibility standards, could you take a look? Thanks!

bradenmacdonald commented:
@cptvitamin Friendly ping re reviewing the proposal above.

cptvitamin commented:
@omarkhan and @bradenmacdonald thanks for putting that proposal together.  I added a few thoughts and questions to it.

omarkhan commented:
@cptvitamin thanks for the review. I have addressed your comments on [the doc](https://docs.google.com/document/d/1VQG2zlL45STxykVBqwGZb1SUkirYPt3iOJBajBmd9ew/edit?usp=sharing)

cptvitamin commented:
@omarkhan i left a bit of feedback. I think you're ready to move forward with the new approach!  Let me know if you have any questions.

jbarciauskas commented:
@bradenmacdonald since Omar has left, what's the next step on this? Should we close this and await a new PR with the updated approach?

bradenmacdonald commented:
@jbarciauskas We'll assign a new developer to take over and then push the new code. Not sure of the timeline yet, but I'd like to get this finished up as soon as possible. If you'd like to use a new PR for the code that uses the new approach, it's fine with me. (Though I'd like to "continue the review where we left off" with the new code and try to avoid a long wait before the review continues, if possible?)

jbarciauskas commented:
Sure, it mainly depends on @cptvitamin 's availability. I don't have a strong feeling about using the same PR or a new one.

bradenmacdonald commented:
@jbarciauskas We are going to abandon this PR for now. There is a lot of work remaining and the site for which we developed this now only needs plain text tooltips, and not rich text.

Note for anyone wanting to add tooltips to a course: The Eucalyptus release includes support for plain text tooltips in HTML/Text content using HTML such as

```html
<span class="clarification" tabindex="0" role="note" aria-label="Clarification">
    <span data-tooltip="(put the tooltip text here)" data-tooltip-show-on-click="true" class="fa fa-info-circle" aria-hidden="true"></span>
    <span class="sr">(also put the tooltip text here)</span>
</span>
```
